### ISSUE TITLE
Expose Low Latency memcache

### ISSUE THREAD

**Issue by LukasReschke** [owncloud/core#13227]
https://github.com/owncloud/core/blob/19b7911d79fca5c88849498298add83bd825565f/lib/private/memcache/factory.php#L64-L75 has a low latence cache which only uses in-memory caches so that no remote hosts will get called.

IMHO it would make sense to expose this as well in the public interface? There are often reasons where it makes sense to store something in a local cache but it wouldn't make sense to store it in a remote cache. (for example because the lookup time might be too long or it is only specific to that instance etc…)

**Comment by LukasReschke** [comment]
@DeepDiver1975 FYI

**Comment by DeepDiver1975** [comment]
> IMHO it would make sense to expose this as well in the public interface?

indeed - good catch - thx

**Comment by LukasReschke** [comment]
Might be sensible to reorder the other cache mechanism then so that Redis et al. are chosen before a local in-memory cachey

**Comment by DeepDiver1975** [comment]
> Might be sensible to reorder the other cache mechanism then so that Redis et al. are chosen before a local in-memory cachey

let's chat Monday about this

**Comment by MorrisJobke** [comment]
@DeepDiver1975 What is the conclusion here?

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 What is the conclusion here?

non yet :see_no_evil:

**Comment by LukasReschke** [comment]
My key points here: Our current caching implementation is dangerous – we cache data in it which is not supposed to be shared such as paths to binaries. Those may differ in distributed deployments on every machine. (a sendmail may be in different directories)

I'd suggest the following changes:

1. Make the low latency cache public
2. Move the priority in the general latency cache to first use distributed memcaches and then local memory caches – currently it's impossible to use e.g. Redis and APCu together since APCu will be used first when it is installed.
3.  Evaluate current usages of the cache and migrate to the low latency cache if useful

To sum up:

- lowLatencyCache
  - APCu
  - APC
  - XCache
  - In-Array Implementation (to be discussed)

- regularCache
  - Redis / Memcache / whatever
  - if not installed: fallback to lowlatencycache

Does that sound reasonable?

**Comment by Xenopathic** [comment]
:+1: :ship: :soon: :question:

:exclamation: :memo: Memcache/Redis :zzz: APCu/XCache :dash:

**Comment by DeepDiver1975** [comment]
> My key points here: Our current caching implementation is dangerous – we cache data in it which is not supposed to be shared such as paths to binaries. Those may differ in distributed deployments on every machine. (a sendmail may be in different directories)

Not that much of an issue - load balanced servers should be setup identical from my pov - but at the end - yes can cause issues

**Comment by DeepDiver1975** [comment]
> Does that sound reasonable?

indeed - go for it if time permits - THX

**Comment by LukasReschke** [comment]
Assigned to @Xenopathic as he volunteers to do that. Thank you very much, Robin!
